### PR TITLE
[WIP] Support different interfaces in addition to sync

### PR DESCRIPTION
**Pull request by jkrems**
The option name is chosen based on how [`wd`](https://www.npmjs.com/package/wd) is calling it.

[Example test](https://github.com/groupon/testium/blob/jk/introduce-driver-type/test/integration/promise-interface.test.coffee):

```coffee
injectBrowser = require 'testium/mocha'
assert = require 'assertive-as-promised'

describe 'driverType: promiseChain', ->
  before injectBrowser driverType: 'promiseChain'

  before ->
    @browser.navigateTo '/'

  it 'can open a page and check the title', ->
    assert.equal 'Test Title', @browser.title()
```

Addresses https://github.com/groupon/testium/issues/164

### CODE REVIEW

**Comment by jkrems** [comment]
Masterplan (WIP):

* Try to separate out the current sync interface from a common core
* Core features: (e.g. `testium-core` eventually)
  - Process management
  - Magic cookie handling (?)
  - Config loading
* Mocha bindings (e.g. `testium-mocha` eventually)
  - Biggest challenge here is to gracefully handle promise vs. sync for getting the screenshot. Otherwise it should be pretty straight-forward.
* Sync driver / browser (e.g. `testium-sync` eventually)
* Promise-chain driver based on `wd` (e.g. `testium-wd`)

`testium` can still expose the same interface, bundling the other libraries. But splitting it up will allow us to have a way to install testium without any native dependencies one day.

The current plan is to just add support for promise chains in addition to the current sync interface. It doesn't look like there will be an easy way to add support for the other `wd` interfaces with static overhead.

cc @EndangeredMassa @ageitgey @khoomeister

File: src/browser/index.coffee [hunk 1]
```diff
@@ -59,6 +59,9 @@ class Browser
     # fixes some phantomjs element size/position reporting
     @setPageSize { height: 768, width: 1024 }
 
+  quit: (callback) ->
```
Comment (EndangeredMassa) [inline src/browser/index.coffee]: Why was this added? Do you just prefer the name `quit`?

**Comment by EndangeredMassa** [comment]
I like this approach.

If we split this out, we'll need to move back to an org again, which is fine. I asked the owner of https://github.com/testium if we could have that again. Their project has been inactive for a long time. We'll see what happens there.

File: src/browser/index.coffee (hunk 1 above)
Comment (jkrems) [inline src/browser/index.coffee]: It was a quick hack because `.quit(cb)` is the same interface that `wd` exposes.

Comment (EndangeredMassa) [inline src/browser/index.coffee]: Ah, sounds fine.

**Comment by EndangeredMassa** [comment]
The owner of that org is not interested in giving up the name.

**Comment by jkrems** [comment]
My goto trick (prior art: [quinnjs](https://github.com/testiumjs) / [buggerjs](https://github.com/buggerjs)) would be to take https://github.com/testiumjs.

**Comment by EndangeredMassa** [comment]
Good idea! I registered just so that we have it.

https://github.com/testiumjs

**Comment by jkrems** [comment]
Some thoughts on how the actual interface between the moving parts could look like (assuming an environment that supports arrow functions and async/await):

Manual linking together of `-wd` and `-core`:

```js
import initTestium from 'testium-core';
import createDriver from 'testium-driver-wd';
import {test} from 'tap';

test('my test', async t => {
  let b = await initTestium().then(createDriver);
  await b.navigateTo('/foo');
  t.equal(await b.getTitle(), 'Hello World', 'shows the right title');
});
```

Same, but `-sync` instead of `-wd`:

```js
import initTestium from 'testium-core';
import createDriver from 'testium-driver-sync';
import {test} from 'tap';

test('my test', async t => {
  let b = await initTestium().then(createDriver);
  b.navigateTo('/foo');
  t.equal(b.getTitle(), 'Hello World');
});
```

Using `-mocha` and explict driver library:

```js
import injectBrowser from 'testium-mocha';
import createDriver from 'testium-driver-sync';
import assert from 'assertive';

describe('my feature', () => {
  before(injectBrowser({ driver: createDriver }));

  before(() => {
    this.browser.navigateTo('/foo');
  });

  it('shows the right title', () => {
    assert.equal('Hello World', this.browser.getTitle());
  });
});
```

Using `-mocha` and automatic driver loading:

```js
import injectBrowser from 'testium-mocha';
import assert from 'assertive';

describe('my feature', () => {
  before(injectBrowser({ driver: 'sync' }));

  before(() => {
    this.browser.navigateTo('/foo');
  });

  it('shows the right title', () => {
    assert.equal('Hello World', this.browser.getTitle());
  });
});
```

Using `-tap` and automatic driver loading:

```js
import test from 'testium-tap';

test('my feature', { driver: 'wd' }, async (t, browser) => {
  await browser.navigateTo('/foo');
  t.equal(await b.getTitle(), 'Hello World', 'shows the right title');
});
```

**Comment by jkrems** [comment]
Thanks to JS's automatic promise flattening, having an async method return the `wd` driver doesn't actually work. Changed the interface of `createDriver` to return `_.extend(testium, { browser })` instead of the browser directly.

**Comment by jkrems** [comment]
Replaced with https://github.com/groupon/testium/pull/171
